crates/wassette/src/lib.rs

```diff
@@ -198,577 +300,241 @@ pub struct ComponentInstance {
-    pub async fn load_component(&self, uri: &str) -> Result<(String, LoadResult)> {
-        debug!(uri, "Loading component");
+    async fn restore_policy_attachment(&self, component_id: &str) -> Result<()> {
+        self.policy_manager.restore_from_disk(component_id).await
+    }
 
-        let downloaded_resource =
+    async fn resolve_component_resource(&self, uri: &str) -> Result<(String, DownloadedResource)> {
+        let resource =
             loader::load_resource::<ComponentResource>(uri, &self.oci_client, &self.http_client)
                 .await?;
+        let id = resource.id()?;
+        Ok((id, resource))
+    }
 
-        // Use optimized loading for manual component loading too
-        let id = downloaded_resource.id()?;
-        let (component, _wasm_bytes) = self.load_component_optimized(downloaded_resource.as_ref(), &id).await
-            .map_err(|e| anyhow::anyhow!("Failed to compile component from path: {}. Error: {}. Please ensure the file is a valid WebAssembly component.", downloaded_resource.as_ref().display(), e))?;
-        // Pre-instantiate the component
-        let instance_pre = self.linker.instantiate_pre(&component)?;
-        let tool_metadata = component_exports_to_tools(&component, &self.engine, true);
-
-        {
-            let mut registry_write = self.registry.write().await;
-            registry_write.unregister_component(&id);
-            registry_write.register_tools(&id, tool_metadata.clone())?;
+    async fn stage_component_artifact(
+        &self,
+        component_id: &str,
+        resource: DownloadedResource,
+    ) -> Result<PathBuf> {
+        let target_path = self.component_path(component_id);
+        match resource {
+            DownloadedResource::Local(path) if path == target_path => Ok(target_path),
+            other => {
+                let _permit = self.storage.acquire_download_permit().await;
+                // Remove any stale component to avoid partial states when copying
+                self.storage
+                    .remove_if_exists(&target_path, "component file", component_id)
+                    .await?;
+                self.storage
+                    .remove_if_exists(
+                        &self.storage.metadata_path(component_id),
+                        "component metadata file",
+                        component_id,
+                    )
+                    .await?;
+                self.storage
+                    .remove_if_exists(
+                        &self.storage.precompiled_path(component_id),
+                        "precompiled component file",
+                        component_id,
+                    )
+                    .await?;
+                other.copy_to(self.storage.root()).await.with_context(|| {
+                    format!(
+                        "Failed to copy component to destination: {}",
+                        self.storage.root().display()
+                    )
+                })?;
```

Comment:
Yep, after looking at this, it feels like these should be internalized to the storage manager part of this, otherwise it is a fairly leaky abstraction